### ISSUE TITLE
Syncing Traffic Lights

### ISSUE THREAD

**Issue by carsakiller** [Starystars67/BeamNG-MP#555]
**Is your feature request related to a problem? Please describe.**
Currently, the traffic lights are not synced. They are controlled in the game in `lua\ge\extensions\core\trafficSignals.lua` by the BeamNG devs with no knowledge of multiplayer or syncing, understandably.

**Describe the solution you'd like**
It could be possible to leave the code the devs have intact and instead add on top of it, calling functions from the module in order to sync traffic lights. Calculating the traffic light states and timings at the server would probably be the most concrete solution, however, the `signals.json` files that describe the different stages and times are found in the game files under each level. This means they are inaccessible at the server. We could pick a “host” player at random to either send the `signals.json` file to the server, or just let their client simulate the lights and send light updates to the server for broadcasting. This would likely lead to more network traffic, though, and also we would be trusting a client to do a lot of the heavy lifting.

**Describe alternatives you've considered**
This could be developed as a plugin instead of becoming part of the core server, but maybe synced traffic lights are part of the “core experience”?

**Additional context**
There is a `resetTimer()` function exposed by the module that resets all the traffic light controller timers. I have already tried simply calling `resetTimer()` across all clients every time a player has finished loading in to the world after joining, which actually works. However, it is very easy to desync again by simply loading in a vehicle and having the clients take a different amount of time to load in the vehicle. Also, every time it is called, the lights may suddenly change state from, say, green to red, instantly. Not very ideal.

It is also possible to get a table of all traffic signal controllers, get their state, and then also set their state. Curious about your ideas/plans if any on syncing traffic lights.

**Comment by Starystars67** [comment]
This is an interesting idea, I too have thought about this topic multiple times.

At this point I think that it could be best to create a server-side plugin that is designed to handle the environment and its syncing.

This could be an "official" plugin but by being a plugin it means that it is as accessible to as many people as possible.

**Comment by OfficialLambdax** [comment]
For this to work a party has to have the authority over the lights, currently the server cant do that as it by itself is just a data relay with a lua engine. And i dont think that a client should ever have this authority. So to actually do this a server side script would have to have knowledge about all the traffic lights any loaded map has and sync it.

Now i never looked at how the traffic lights are setup in the game, but there are likely references in the main mission files of a map that a server side script could auto extract (or already provide for the vanilla maps) and then handle.

However im not quite sure if there is really an interest in the wider community for traffic light sync. If there is i have no problem putting this together

**Comment by WiserTixx** [comment]
I think it would be best if the server had some kind of clock that it syncs with the clients, so that the client can then sync time based things like traffic lights while keeping the protocol implementation flexible enough so that other time based things can also be synced in the future.

**Comment by OfficialLambdax** [comment]
So your thinking of making things like eg traffic lights so deterministic that all youd have to have is a numeric value (eg. server life time) that the server sends to each client and this num can then be used to eval teh state of a traffic light?

Eg time 1000 would always relate to X traffic light to be Red. Is that what you mean?

**Comment by WiserTixx** [comment]
Yes, with adequate compensation for latency. However it might be better to implement this in the new protocol instead.

**Comment by StanleyDudek** [comment]
Sorry to dig this up, but I just wanted to provide insight for @carsakiller as well as document a simple approach to this, so simple, yes it could be built into the server perhaps, such that the uptime of the server could simply serve as the main sync clock for things like this.

In effect, the following are all that's needed to have the server be the time authority, regularly tell clients what the time is, and by virtue of the signal light cycles staying on timer, the lights states are synchronized across clients.

Something more elegant would be implemented by a server owner, but I found this solution myself working on freeRoamMP, so that traffic accidents across clients would happen less (:

server:
```lua
...
local signalTimer = MP.CreateTimer()
...
MP.RegisterEvent("trafficLightTimer","trafficLightTimer")
MP.CreateEventTimer("trafficLightTimer", 10000)
...
function trafficLightTimer()
    if #MP.GetPlayers() >= 1 then
        MP.TriggerClientEvent(-1, "rxTrafficSignalTimer", tostring(signalTimer:GetCurrent()))
    end
end
...
```
client:
```lua
...
local function rxTrafficSignalTimer(data)
    core_trafficSignals.setTimer(tonumber(data))
end
...
AddEventHandler("rxTrafficSignalTimer", rxTrafficSignalTimer)
...
```

**Comment by turbskiiii** [comment]
Is this going to implemented anytime soon?

**Comment by StanleyDudek** [comment]
> Is this going to implemented anytime soon?

I don't think there are any plans for this to be an official part of BeamMP anytime soon, but I did create this: https://github.com/StanleyDudek/trafficSignalSync

which a server owner can use to accomplish the same

**Comment by turbskiiii** [comment]
> > Is this going to implemented anytime soon?
> 
> I don't think there are any plans for this to be an official part of BeamMP anytime soon, but I did create this: https://github.com/StanleyDudek/trafficSignalSync
> 
> which a server owner can use to accomplish the same

Now does this only work for certain maps or?

**Comment by StanleyDudek** [comment]
It works for the stock maps that have traffic signal systems, and should in theory work for any modded map that uses the same base game traffic signal systems; if a modded map has traffic signals but does not use the base game system, or uses a custom system for controlling timing, it may not work.

**Comment by turbskiiii** [comment]
> It works for the stock maps that have traffic signal systems, and should in theory work for any modded map that uses the same base game traffic signal systems; if a modded map has traffic signals but does not use the base game system, or uses a custom system for controlling timing, it may not work.

I'll look into it will update you!
